exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -92,38 +94,110 @@ public DrillFunctionRegistry(ScanResult classpathScan) {
+          aggregateFunctions.put(paramCount, func);
+        } else {
+          final Pair<Integer, Integer> argNumerRange = getArgNumerRange(name, func);
+          functions.put(argNumerRange, func);
         }
+
+        if(!func.isDeterministic()) {
+          isDeterministic = false;
+        }
+      }
+      for (Entry<Pair<Integer, Integer>, Collection<DrillFuncHolder>> entry : functions.asMap().entrySet()) {
+        final DrillSqlOperator drillSqlOperator;
+        final Pair<Integer, Integer> range = entry.getKey();
+        final int max = range.getRight();
+        final int min = range.getLeft();
+        drillSqlOperator = new DrillSqlOperator(
+            name,
+            Lists.newArrayList(entry.getValue()),
+            min,
+            max,
+            isDeterministic);
+        operatorTable.add(name, drillSqlOperator);
+      }
+      for (Entry<Integer, Collection<DrillFuncHolder>> entry : aggregateFunctions.asMap().entrySet()) {
+        operatorTable.add(name, new DrillSqlAggOperator(name, Lists.newArrayList(entry.getValue()), entry.getKey()));
       }
     }
+
+    registerCalcitePlaceHolderFunction(operatorTable);
+  }
+
+  /**
+   * These {@link DrillSqlOperator} merely act as a placeholder so that Calcite
+   * allows convert_to(), convert_from(), flatten(), date_part() functions in SQL.
+   */
+  private void registerCalcitePlaceHolderFunction(DrillOperatorTable operatorTable) {
+    final String convert_to = "CONVERT_TO";
+    final String convert_from = "CONVERT_FROM";
+    final String flatten = "FLATTEN";
+    final String date_part = "DATE_PART";
+
+    operatorTable.add(convert_to,
+        new DrillSqlOperator(convert_to,
+            2,
+            true));
+    operatorTable.add(convert_from,
+        new DrillSqlOperator(convert_from,
+            2,
+            true));
+    operatorTable.add(flatten,
+        new DrillSqlOperator(flatten,
+            1,
+            true));
+    operatorTable.add(date_part,
+        new DrillSqlOperator(date_part,
+            2,
+            true));
   }
 
+  private Pair<Integer, Integer> getArgNumerRange(final String name, final DrillFuncHolder func) {
+    switch(name.toUpperCase()) {
+      case "CONCAT":
+        return Pair.of(1, Integer.MAX_VALUE);
+
+      // Drill does not have a FunctionTemplate for the lpad/rpad with two arguments.
+      // It relies on DrillOptiq.java to add a third dummy argument to be acceptable
+      // by the FunctionTemplate in StringFunctions.java
+      case "LPAD":
+      case "RPAD":
+        return Pair.of(2, 3);
+
+      // Similar to the reason above, DrillOptiq.java is used for rewritting
+      case "LTRIM":
+      case "RTRIM":
+      case "BTRIM":
+        return Pair.of(1, 2);
+
+      // Similar to the reason above, DrillOptiq.java is used for rewritting
+      case "LENGTH":
+        return Pair.of(1, 2);
+
+      default:
+        return Pair.of(func.getParamCount(), func.getParamCount());
+    }
+  }
```

Comment:
@jacques-n , here is my understanding about the dummy functions @hsuanyi is talking about.

For 'lpad' function, Calcite allows both 2 arguments or 3 arguments. Drill has only implementation for the case of 3 arguments. What Drill did is DrillOptiq will rewrite the case of 2 arguments into 3 arguments [2]. The dummy function is for the case of 2 arguments, since Drill's run-time generate logic does not need the implementation of 2 arguments at all, due to the rewrite in DrillOptiq. 

Another option is 1) to remove the rewrite in DrillOptiq and 2) add the implementation for the case of 2 arguments (which is essentially same as the one for 3 arguments).  The new implementation for 2 arguments case is not dummy at all. 

  

[1] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/StringFunctions.java#L702-L707
[2] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillOptiq.java#L429-L434